framework/kms-keyring.md

```diff
@@ -309,7 +312,84 @@ If the response is successfully verified, OnDecrypt MUST do the following with t
 If OnDecrypt fails to successfully decrypt any [encrypted data key](#structures.md#encrypted-data-key),
 then OnDecrypt MUST output the unmodified input [decryption materials](#structures.md#decryption-materials).
 
+## Configuration Intent
+
+### OnEncrypt Goal
+
+When a user configures a KMS keyring with key IDs
+and use that keyring to encrypt a message,
+they are stating their intent that they need each one of those CMKs to be able to
+independently decrypt the resulting encrypted message.
+
+For example, if a user configures a KMS keyring with CMKs A, B, and C
+and uses that keyring to encrypt a message,
+then that user expects three things to be true:
+
+1. CMK A can decrypt the encrypted message.
+1. CMK B can decrypt the encrypted message.
+1. CMK C can decrypt the encrypted message.
+
+If any of these are not true of the resulting encrypted message,
+the keyring has failed to honor the user's intent.
+
+In order to accomplish this,
+the KMS keyring MUST contribute three encrypted data keys to the encryption materials:
+one from CMK A, one from CMK B, and one from CMK C.
+
+### OnDecrypt Goal
+
+When a user configures a KMS keyring for use on decrypt,
+they are stating their intent for which CMKs
+the keyring will *attempt* to use to decrypt encrypted data keys.
+
+For example, if a user configures a KMS keyring with CMK C
+and uses it to decrypt an encrypted message
+that contains encrypted data keys for CMKs A, B, and C,
+then the keyring will attempt to decrypt using CMK C.
+
+However, if the keyring attempts to decrypt using CMK C and cannot,
+this failure still honors the configured intent and MUST NOT halt decryption.
+The configured intent is that they keyring MUST *attempt* with these CMKs,
+not that they MUST *succeed*.
+
+### Why OnEncrypt and OnDecrypt are different
+
+On encrypt, the user describes their intent
+for the requirements to decrypt the resulting message.
+Because of this,
+the keyring MUST create encryption materials that satisfies those requirements.
+
+On decrypt, the user provides resources that *attempt* to do that decryption.
+
+This is an asymmetric relationship with very different implications on failure.
+If the keyring encounters a problem on encrypt,
+it cannot fully honor the decryption requirements and so MUST halt message encryption.
+
+However, on decrypt the keyring is not creating anything.
+It is instead attempting to satisfy the requirements that were set on encryption.
+If the keyring cannot satisfy those requirements it MUST NOT halt message decryption.
+
+No keyring can know if it is the last keyring to attempt decryption.
+If all keyrings are exhausted and none of them were able to decrypt an encrypted data key
+then the cryptographic materials manager that managed those keyrings
+will halt message decryption.
+(See [Default Cryptographic Materials Manager](./default-cmm.md))
+
+### Requirements
```

Comment:
I like being able to state high level intentions like this, but I think it might be a bit confusing or miss-able underneath all of this soft discussion of intention. These requirements are invariants which the above specification should cover. i.e. If your implementation follows the above specification for behavior of this keyring, it will have these properties. These invariants are stated elsewhere to clarify intentions and highlight important properties that the above spec offers. I'm not quite sure how to get this across, but "Requirements" doesn't seem like the right title.

Reply:
I think I agree. Perhaps examples? I know that clarity can be a form of repetition, but I think that the encrypt requirement is special here, and the decrypt requirement just flows from the keyring intent.

I also wonder if some of this intent language should not also be moved to the Multi-Keyring as these two share a lot of intent...

Reply:
> If your implementation follows the above specification for behavior of this keyring, it will have these properties

Agreed, but I see this as almost a parity check for that: this is a very concise, straight-forward, statement of the top-level requirements. `Operation` is a description of *how* to achieve those requirements and `Configuration Intent` is an explanation of *why* we have those requirements.

I agree that there might be a better place to put this. I'm not entirely sure where in the doc that is, but I'll play with some options. For a starting point, I wanted to make sure it was in there at all. :)

> I also wonder if some of this intent language should not also be moved to the Multi-Keyring as these two share a lot of intent...

Not moved, but replicated. I've created #76 to track that.